### DISCUSSION TITLE
Navigate overwrite context value from index page

### DISCUSSION THREAD

**Discussion by popuguytheparrot** [vercel/next.js#46044] [Help]
Hi

I'm using gsp to create a value for a provider at the page level. After the build, I need the value in the provider to always be the same. How to achieve this?

How can you inline a value in a provider and never change it again?

**Comment by icyJoseph** [comment]
I guess you use the `pageProps` and pass it down to a `<Context.Provider value={pageProps.foo}>{children}</Context.Provider>`, sort of right?

I think there's two ways....

Merge the new into the current, `{...new,...current}`, so that the old prevails during merge. So that if `foo` is undefined on further navigation, you'd still keep your context value.

And, well, going against everything I know to be true, good, and best practice, what you can do is "freeze" the initial pageProps, by passing it to `useState`, never update it, and use the value (not the setter) of `useState` as context provider value. Not even gonna post code for that, but I think you probably understand what I mean? Of course, this introduces tons of potential issues and bugs... so be careful, I'd say merging is a bit more friendly.

**Reply by icyJoseph** [reply]
Thought I'd ask, why though? Feels wrong to couple other pages to the props of another page.

**Reply by popuguytheparrot** [reply]
This approach will work if the user visits the main page

If the user enters the application through any other, then this state will not be

I need a way that would allow me to pull out the data that is in
_next/build_id/data/index.json

The only option that I see is to somehow request data in _document and paste them as <script type=json>

**Reply by popuguytheparrot** [reply]
The idea is that the index.tsx page in gsp will fetch common data for header, footer, sidebar and pass to context

**Reply by popuguytheparrot** [reply]
in the build time I make a request and share data through the context for the entire application.

The entry point will be index.tsx page for fetch

**Reply by icyJoseph** [reply]
Right so you are trying to solve the global data problem. 

That's not trivial at all. The most common solution is using `getInitialProps` at _app level, but that has issues too. 

Pretty much you gotta handle the client calls to getInitialProps props too, but it is doable, just has trade offs. 

Mainly that only pages that implement getStaticProps will be pre-rendered.

I had a hunch this was the root issue. 

Otherwise you'll end up buried in hacks and bugs.
